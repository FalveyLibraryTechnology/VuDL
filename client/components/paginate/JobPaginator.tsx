import React, { useEffect } from "react";
import PropTypes from "prop-types";
import JobPaginatorZoomToggle from "./JobPaginatorZoomToggle";
import PaginatorControls from "./PaginatorControls";
import PaginatorList from "./PaginatorList";
import { usePaginatorContext } from "../../context/PaginatorContext";

const JobPaginator = ({ initialCategory, initialJob }) => {
    const {
        state: { category, job },
        action: { loadJob },
    } = usePaginatorContext();

    useEffect(() => {
        loadJob(initialCategory, initialJob);
    }, []);

    return (
        <div id="paginator">
            <div className="row">
                <div className="six col">
                    <JobPaginatorZoomToggle />
                </div>
                <div className="six col">
                    <p>
                        {category} &gt; {job}
                    </p>
                    <PaginatorControls />
                    <PaginatorList />
                </div>
            </div>
        </div>
    );
};

JobPaginator.propTypes = {
    initialCategory: PropTypes.string,
    initialJob: PropTypes.string,
};

export default JobPaginator;
