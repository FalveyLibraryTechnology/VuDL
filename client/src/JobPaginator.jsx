import React, { useEffect } from "react";
import PropTypes from "prop-types";

import JobPaginatorZoomToggle from "./JobPaginatorZoomToggle";
import PaginatorControls from "./PaginatorControls";
import PaginatorList from "./PaginatorList";
import useJobPaginator from "./useJobPaginator";

const JobPaginator = ({ initialCategory, initialJob }) => {
    const {
        state: { currentPage, zoom, order, category, job },
        action: {
            setLabel,
            getLabel,
            getMagicLabel,
            setPage,
            nextPage,
            prevPage,
            deletePage,
            save,
            loadJob,
            autonumberFollowingPages,
            getJobImageUrl,
            setZoom,
        },
    } = useJobPaginator(initialCategory, initialJob);

    useEffect(() => {
        loadJob();
    }, []);

    return (
        <div id="paginator">
            <div className="row">
                <div className="six col">
                    <JobPaginatorZoomToggle
                        enabled={order.length > 0}
                        zoom={zoom}
                        getJobImageUrl={(size) => getJobImageUrl(order[currentPage], size)}
                    />
                </div>
                <div className="six col">
                    <p>
                        {category} &gt; {job}
                    </p>
                    <PaginatorControls
                        autonumberFollowingPages={autonumberFollowingPages}
                        currentPage={currentPage}
                        getLabel={getLabel}
                        getMagicLabel={getMagicLabel}
                        setLabel={setLabel}
                        prevPage={prevPage}
                        nextPage={nextPage}
                        deletePage={deletePage}
                        save={save}
                        toggleZoom={() => {
                            setZoom(!zoom);
                        }}
                        zoom={zoom}
                        pageCount={order.length}
                    />
                    <PaginatorList
                        setPage={setPage}
                        getLabel={getLabel}
                        getMagicLabel={getMagicLabel}
                        setLabel={setLabel}
                        getJobImageUrl={(imageNumber, size) => getJobImageUrl(order[imageNumber], size)}
                        currentPage={currentPage}
                        pageCount={order.length}
                    />
                </div>
            </div>
        </div>
    );
};

JobPaginator.propTypes = {
    // VuDLPrep
    app: PropTypes.shape({
        activateJobSelector: PropTypes.func,
        ajax: PropTypes.func,
        getJobImageUrl: PropTypes.func,
        getJobUrl: PropTypes.func,
        fetchJSON: PropTypes.func,
    }),
    initialCategory: PropTypes.string,
    initialJob: PropTypes.string,
};

export default JobPaginator;
