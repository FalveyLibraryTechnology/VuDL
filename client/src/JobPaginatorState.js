const confirmSavedMagicLabels = (count) => {
    return window.confirm("You will be saving " + count + " unreviewed, auto-generated label(s). Are you sure?");
};

const findNewPagePosition = (page, list) => {
    const foundIndex = list.findIndex((listItem) => {
        return listItem.filename >= page;
    });
    return foundIndex === -1 ? list.length : foundIndex;
};

const deletePageValidation = (order) => {
    if (order.length < 2) {
        alert("You cannot delete the last page in a job.");
        return false;
    }
    return window.confirm("Are you sure you wish to delete the current page?");
};

const getAddedPages = (order, pages) => {
    return pages.reduce((order, page) => {
        order.splice(findNewPagePosition(page, order), 0, {
            filename: page,
            label: null,
        });
        return order;
    }, order);
};

const getNonRemovedPages = (order, deletedPages) =>
    order.filter((o) =>
        deletedPages.every((page) => {
            return o.filename !== page;
        })
    );

const getLabel = (order, imageNumber) => {
    return typeof order[imageNumber] === "undefined" ? null : order[imageNumber].label;
};

const countMagicLabels = (order, startAt) => {
    let count = 0;
    for (let startIndex = startAt; startIndex < order.length; startIndex++) {
        if (null === getLabel(order, startIndex)) {
            count++;
        }
    }
    return count;
};

const userMustReviewMoreLabels = (order) => {
    const count = countMagicLabels(order, 0);
    return count > 0 && !confirmSavedMagicLabels(count);
};

export {
    countMagicLabels,
    deletePageValidation,
    getAddedPages,
    getNonRemovedPages,
    getLabel,
    confirmSavedMagicLabels,
    findNewPagePosition,
    userMustReviewMoreLabels,
};
