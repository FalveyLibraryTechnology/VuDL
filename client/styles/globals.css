:root,
[color-scheme="light"] {
    --body-bg: #fafbfc;
    --heading-color: #4d4d4d;
    --link-color: #008aed;

    --btn-bg: #485a6c;
    --btn-text: #ffffff;
    --btn-border: #275575;
    --btn-focus-bg: #3498db;

    --btn-primary-bg: #27ae60;
    --btn-primary-text: #ffffff;
    --btn-primary-border: #275575;
    --btn-primary-focus-bg: #155724;

    --btn-danger-bg: red;
    --btn-danger-text: #ffffff;
    --btn-danger-border: #a10705;
    --btn-danger-focus-bg: #a10705;

    --btn-logout-bg: #1abc9c;
    --btn-logout-text: #ffffff;
    --btn-logout-border: #16a085;
    --btn-logout-focus-bg: #1abc9c;

    --btn-toggle-bg: #7f8c8d;
    --btn-toggle-text: #ffffff;
    --btn-toggle-border: #34495e;
    --btn-toggle-focus-bg: #53626f;

    --group-border: #53626f;

    --input-border: #6a737d;
    --input-focus-border: #2188ff;

    --label-hover: #c8e1ff;

    --page-list-bg: #ffffff;
    --page-list-border: #c1c1c1;

    --thumbnail-bg: #e1e1e1;
    --thumbnail-focus-bg: #51aded;
    --thumbnail-number-bg: #ffffff;
    --thumbnail-number-text: #000000;
    --thumbnail-preview-border: #000000;

    --tree-item-focus-bg: #000000;
    --tree-item-focus-text: #ffffff;

    --shadow-gray: #c8e1ff;
}

[color-scheme="dark"] {
}
