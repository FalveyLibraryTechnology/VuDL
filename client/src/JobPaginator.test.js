import React from "react";
import { act } from "react-dom/test-utils";
import { beforeEach, describe, expect, it, jest } from "@jest/globals";
import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import JobPaginator from "./JobPaginator";
import * as JobPaginatorState from "./JobPaginatorState";

jest.mock("./JobPaginatorState");
jest.mock("./PaginatorControls", () => () => "PaginatorControls");
jest.mock("./PaginatorList", () => () => "PaginatorList");
jest.mock("./JobPaginatorZoomToggle", () => () => "JobPaginatorZoomToggle");

describe("JobPaginator", () => {
    let props;
    beforeEach(() => {
        jest.spyOn(JobPaginatorState, "getJob").mockResolvedValue({
            order: [
                {
                    filename: "test1",
                    label: "test2",
                },
            ],
        });

        jest.spyOn(JobPaginatorState, "getStatus").mockResolvedValue({
            file_problems: {
                deleted: [],
                added: [],
            },
        });
        props = {
            initialCategory: "testCategory",
            initialJob: "testJob",
        };
    });

    it("renders", () => {
        const wrapper = shallow(<JobPaginator {...props} />);
        expect(wrapper.contains(props.initialCategory)).toBeTruthy();
        expect(wrapper.contains(props.initialJob)).toBeTruthy();
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("should loadJob from useEffect", async () => {
        await act(async () => {
            await mount(<JobPaginator {...props} />);
        });
        expect(JobPaginatorState.getJob).toHaveBeenCalled();
        expect(JobPaginatorState.getStatus).toHaveBeenCalled();
    });
});
