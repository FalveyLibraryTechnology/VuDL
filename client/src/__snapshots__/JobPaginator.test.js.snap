// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobPaginator renders 1`] = `
<div
  id="paginator"
>
  <div
    class="row"
  >
    <div
      class="six col"
    >
      JobPaginatorZoomToggle
    </div>
    <div
      class="six col"
    >
      <p>
        testCategory &gt; testJob
      </p>
      PaginatorControlsPaginatorList
    </div>
  </div>
</div>
`;
